File: 2008Practise/PracticeProblems/AlwaysTurnLeft.py
def alwaysTurnLeft ():
    nCases = 0
    l = 0
    x = [0,0]
    y = [0,0]
    f = open ("c:\\gcj\\2008Practise\\PracticeProblems\\B-large-practice.in",'r')
    g = open ("c:\\gcj\\2008Practise\\PracticeProblems\\B-large-practice.out",'w')
    lines = f.readlines()
    nCases = int(lines[0])
    types = {'[1, 0, 0, 0]' : '1', '[0, 1, 0, 0]' : '2', '[1, 1, 0, 0]' : '3', '[0, 0, 1, 0]' : '4',
             '[1, 0, 1, 0]' : '5', '[0, 1, 1, 0]' : '6', '[1, 1, 1, 0]' : '7', '[0, 0, 0, 1]' : '8',
             '[1, 0, 0, 1]' : '9', '[0, 1, 0, 1]' : 'a', '[1, 1, 0, 1]' : 'b', '[0, 0, 1, 1]' : 'c',
             '[1, 0, 1, 1]' : 'd', '[0, 1, 1, 1]' : 'e', '[1, 1, 1, 1]' : 'f'}
    lines.pop(0)
    case = 0
    for line in lines:
        case = case +1
        g.write("Case #" + str(case) + ":\n")
        cells =  getCells(line)
        sortedCells = sorted(cells, key=lambda x: (x[0][1], x[0][0]))

        row = 0
        row_text = ''
        for cell in sortedCells:
            if (cell[0][1] == row):
                cell.pop(0)
                row_text = row_text + types[str(cell)]
            else:
                cell.pop(0)
                g.write(row_text+'\n')
                row = row+1
                row_text = types[str(cell)]

        g.write(row_text+'\n')





    return 1


def stepPermute(direction,step):
    ret = direction
# direction: [0, 1] = 's', [0,-1] = 'n', [1, 0] = 'e', [-1,0] = 'w'
    if(step == 'L'):
        if (direction == [0, 1]):
            ret = [1, 0]
        elif(direction == [0,-1]):
            ret = [-1, 0]
        elif(direction == [1,0]):
            ret = [0, -1]
        else:
            ret = [0, 1]
    elif(step == 'R'):
        if (direction == [0, 1]):
            ret = [-1, 0]
        elif(direction == [0,-1]):
            ret = [1, 0]
        elif(direction == [1,0]):
            ret = [0, 1]
        else:
            ret = [0, -1]
    return ret



def getCells(walks):
    walkss = walks.split()
    cells = []
    direction = [0, -1]
    position = [0,-1]


    for walk in walkss:
        direction = map(lambda x: x * -1, direction)
#        print direction
        steps = list(walk)
        for step in steps:
            if(step == 'W'):
                position = [position[0] + direction[0], position[1] + direction[1]]
                cells = addToCell(cells,position,direction,step)
            else:
                cells = addToCell(cells,position,direction,step)
                direction = stepPermute(direction,step)

        cells.pop()


    return cells

def addToCell(cells,position,direction,step):
# cells(position, can-go-north, south, west, east)
    if (not position in [x[0] for x in cells]):
        cells.append([position,0,0,0,0])

    cell = [x for x in cells if (x[0] == position)]
    cell = cell[0]
#    print cell
#    print step
#    print position
#    print direction

    if(step == 'W'):
        if (direction == [1, 0]):
            cell[3] = 1
        elif (direction == [-1, 0]):
            cell[4] = 1
        elif (direction == [0, 1]):
            cell[1] = 1
        else:
            cell[2] = 1
#    elif (step == 'L'):
#        if (direction == [1, 0]):
#            cell[4] = 0
 #       elif (direction == [-1, 0]):
 #           cell[3] = 0
 #       elif (direction == [0, 1]):
 #           cell[2] = 0
 #       else:
 #           cell[1] = 0
 #   else:
 #       if (direction == [1, 0]):
 #           cell[1] = 0
 #           cell[4] = 0
 #       elif (direction == [-1, 0]):
 #           cell[2] = 0
 #           cell[3] = 0
 #       elif (direction == [0, 1]):
 ##           cell[4] = 0
 #           cell[2] = 0
 ##       else:
  #          cell[3] = 0
  #          cell[1] = 0
#    print cells
    return cells

alwaysTurnLeft()

def test():
    m = [[1, 0],1,0,0,1]
    m.pop(0)
    return str(m)

test()

File: 2008Practise/PracticeProblems/AlwaysTurnLeft.clj
(ns alwaysTurnLeft
  (:use [clojure.java.io  :only [reader writer]]
        [clojure.string  :only [split]])
  (:import (java.io File)))

(def infile "c:\\gcj\\2008Practise\\PracticeProblems\\B-large-practice.in")
(def outfile "c:\\gcj\\2008Practise\\PracticeProblems\\B-large-practice.out.clj")
;(def infile "/home/andreask/gcj/2008Practise/PracticeProblems/B-small-practice.in")
;(def outfile "/home/andreask/gcj/2008Practise/PracticeProblems/B-small-practice.out.clj")


(defn read-lines [file]
  (with-open [r (reader file)]
    (doall (line-seq r))))

(defn write-lines
	  "Writes lines (a seq) to f, separated by newlines.  f is opened with
	  writer, and automatically closed at the end of the sequence."
	  [f lines]
	  (with-open [writer (writer f)]
	    (loop [lines lines]
	      (when-let [line (first lines)]
	        (.write writer (str line))
	        (.newLine writer)
	        (recur (rest lines))))))




(defn step-permute [dir step]
  (cond
   (= step \L)
     (cond
      (= dir [0, 1]) [1, 0]
      (= dir [0, -1]) [-1, 0]
      (= dir [1, 0]) [0, -1]
      true [0, 1])
   (= step \R)
     (cond
      (= dir [0, 1]) [-1, 0]
      (= dir [0, -1]) [1, 0]
      (= dir [1, 0]) [0, 1]
      true [0, -1])
   true dir))


(defn add-to-cell [c p d step]
  (let [cells-out (if (not (some #(= % p) (map first c))) (vec (conj c [p 0 0 0 0])) (vec c))
        cell (first (filter (fn [x] (= p (first x))) cells-out))
        cell-pos (.indexOf (map first cells-out) p)]
    (do
;      (print "in add to")
;      (print c)
    (if (= step \W)

      (cond
       (= [1, 0] d) (assoc cells-out cell-pos (assoc cell 3 1))
       (= [-1, 0] d) (assoc cells-out cell-pos (assoc cell 4 1))
       (= [0, 1] d) (assoc cells-out cell-pos (assoc cell 1 1))
       true (assoc cells-out cell-pos (assoc cell 2 1)))
      cells-out))))



(defn step-done [steps pos dir cells-out]
  (do
;    (print "in step done")
  (loop [s 0 p pos d dir c cells-out]
    (if (= s (count steps))
      (do
;        (print "fins")
      [(butlast c) d p])
      (do
;        (print (nth steps s))
      (if (= (nth steps s) \W)
        (let [new-pos [(+ (first p)  (first d)) (+ (second p) (second d))]]
          (recur (inc s) new-pos d (add-to-cell c new-pos d (nth steps s))))
        (recur (inc s) p (step-permute d (nth steps s)) (add-to-cell c p d (nth steps s)))))))))









(defn get-cells [walks]
  (let [walkss (split walks #" ")
        cells []
        direction  [0, -1]
        position [0, -1]]
    (loop [w 0 cells-out cells dir direction pos position]

      (if (= w (count walkss))
        cells-out
        (let [direct (map #(* % -1) dir)
              steps (into [] (nth walkss w))
              step-res (step-done steps pos direct cells-out)]
           (do
;             (print steps)
          (recur (inc w) (first step-res) (second step-res) (nth step-res 2))))))))



(defn always-turn-left []
  (let [lines (read-lines infile)
        nCases (read-string (first lines))
        types {"[1 0 0 0]" "1", "[0 1 0 0]"   "2", "[1 1 0 0]"   "3", "[0 0 1 0]"   "4",
             "[1 0 1 0]"   "5", "[0 1 1 0]"   "6", "[1 1 1 0]"   "7", "[0 0 0 1]"   "8",
             "[1 0 0 1]"   "9", "[0 1 0 1]"   "a", "[1 1 0 1]"   "b", "[0 0 1 1]"   "c",
             "[1 0 1 1]"   "d", "[0 1 1 1]"   "e", "[1 1 1 1]"   "f"}
        out-file outfile]

    (with-open [writer (writer out-file)]
      (loop [case 1]
      (if (= case (inc nCases))
          "Done"
          (do
            (.write writer (str "Case #" case ":"))
            (.newLine writer)
            (loop [row 0]
              (let [sort-cells (sort-by (juxt (fn [x] (second (first x))) (fn [x] (first (first x)))) (get-cells (nth lines case)))
                    ress (filter (fn [x] (= row (second (first x)))) sort-cells)]
              (if (= 0 (count ress))
               "Done"
                (let [ln (reduce str (map (fn [x] (types (str (vec (rest x))))) ress))]
                  (do
                    (.write writer (str ln))
                     (.newLine writer)
                (recur (inc row)))))))
            (recur (inc case))))))))
;    (loop [l 1 cells []]
;      (if (= l (inc (count lines)))
 ;       "Done"
;        (let [cells (get-cells (nth lines l))
;              sorted-cells (sort-by (juxt (fn [x] (second (first x))) (fn [x] (first (first x)))) cells)]
;          cells)))))

(always-turn-left)

(str (vec (rest [[0 1] 0 7])))

;(conj )
;(replace [0 2 77] (list 0 1 2 3 4))


